src/assisted_installer_controller/assisted_installer_controller.go

```diff
@@ -766,151 +754,52 @@ func (c controller) updatePendingOLMOperators() error {
 }
 
 // waitForOLMOperators wait until all OLM monitored operators are available or failed.
-func (c controller) waitForOLMOperators() bool {
-	c.log.Infof("Checking OLM operators")
-	operators, _ := c.getProgressingOLMOperators()
+func (c controller) waitForOLMOperators(ctx context.Context) error {
+	operators, err := c.getProgressingOLMOperators()
+	if err != nil {
+		return err
+	}
 	if len(operators) == 0 {
-		return true
+		return nil
 	}
-	for _, operator := range operators {
-		csvName, err := c.kc.GetCSVFromSubscription(operator.Namespace, operator.SubscriptionName)
-		if err != nil {
-			c.log.WithError(err).Warnf("Failed to get subscription of operator %s", operator.Name)
-			continue
-		}
 
-		csv, err := c.kc.GetCSV(operator.Namespace, csvName)
-		if err != nil {
-			c.log.WithError(err).Warnf("Failed to get %s", operator.Name)
-			continue
-		}
+	handlers := make([]*ClusterServiceVersionHandler, 0)
 
-		operatorStatus := utils.CsvStatusToOperatorStatus(string(csv.Status.Phase))
+	for index := range operators {
+		handlers = append(handlers, NewClusterServiceVersionHandler(c.kc, &operators[index]))
+	}
 
-		// FIXME: We retry the check of the operator status in case it's in failed state to WA bug 1968606
-		// Remove this code when bug 1968606 is fixed
-		if utils.IsStatusFailed(operatorStatus) && c.retryMap[operator.Name] < failedOperatorRetry {
-			c.retryMap[operator.Name]++
-			c.log.Warnf("Operator %s has failed state retry(%d/%d) the check.", operator.Name, c.retryMap[operator.Name], failedOperatorRetry)
-			continue
+	areOLMOperatorsAvailable := func() bool {
+		if len(handlers) == 0 {
+			return true
 		}
 
-		if operator.Status != operatorStatus || (operator.StatusInfo != csv.Status.Message && csv.Status.Message != "") {
-			c.log.Infof("CSV %s updated, status: %s -> %s, message: %s -> %s.", operator.Name, operator.Status, operatorStatus, operator.StatusInfo, csv.Status.Message)
+		currentHandlers := handlers[:]
 
-			err = c.ic.UpdateClusterOperator(context.TODO(), c.ClusterID, operator.Name, operatorStatus, csv.Status.Message)
-			if err != nil {
-				c.log.WithError(err).Warnf("Failed to update olm %s status", operator.Name)
-				continue
+		for index := range currentHandlers {
+			if c.isOperatorAvailable(currentHandlers[index]) {
+				handlers = append(handlers[:index], handlers[index+1:]...)
```

Comment:
This seems buggy.
Looks like you remove a handler from `handlers` but iterate over `currentHandlers`.
After c.isOperatorAvailable(currentHandlers[index]) is true once the rest of the loop is out of sync.

Reply:
You're right. fixed.